### ISSUE TITLE
Password dependent authentication for workers

### ISSUE THREAD

**Issue by peregrineshahin** [glinscott/fishtest#1724]
Current authentication for WEB users has a login API that only validates passwords once and sends back a CSRF token to be used for any API that requires the authentication/privilege of the user.

Now, for workers, they do not depend on a login API thus they don't get back a CSRF-token This necessitates the validation of each request via sending the password with each request made to the server, the downside of that is if passwords are hashed in the database soon, then each request will require the hashing of the submitted password again and again and again which will be the culprit of very high load on the server.

The correct solution is to solve this like it is solved for users on web.
1- log in the worker
2- generate a CSRF-token for the user
3- send it back in the request of the login to the worker
4- only validate API requests via the CSRF token.

**Comment by vondele** [comment]
is the current pending PR not using a cache to avoid this high load? 

This doesn't mean that this is not a valid suggestion, but the load issue might be non-existent.

**Comment by vdbergh** [comment]
Just thinking theoretically.

I am trying to figure out how pyramid does this. I have the impression that it does not set any state on the server. As I understand it the client gets a cookie, signed by the server, that it is logged in with a specific username. This cookie then serves as the client credentials for a certain time. Is this correct?

To do the signing the server needs a secret. I assume this is generated at server start up? 

Such a system would not be terribly hard to implement for the worker (or perhaps it is possible to use the same mechanism that is used for views?).

**Comment by vdbergh** [comment]
So when the worker does an api call with username/password, in its reply the server should return a timestamp and a signature of the username/timestamp/ipaddress, both of which are stored in the worker info dictionary.

The next api call the server can use this signature to see if the username/timestamp supplied by the user can be trusted.

**Comment by peregrineshahin** [comment]
> So when the worker does an api call with username/password, in its reply the server should return a timestamp and a signature of the username/timestamp/ipaddress, both of which are stored in the worker info dictionary.
> 
> The next api call the server can use this signature to see if the username/timestamp supplied by the user can be trusted.

Not really, the worker should only just login, gets back a token and store this token in RAM, i.e. inside a class variable and then starts to send normal api calls with that token.
The server can just authenticate that via requiring a csrf token.
just exactly like the view settings.

**Comment by vdbergh** [comment]
Does your system set state on the server?

**Comment by peregrineshahin** [comment]
What do you mean it does not save state on the server, then how we authenticate web views?
We already have a session management, I assume they are saved here

```
    """This function returns a Pyramid WSGI application."""
    session_factory = SignedCookieSessionFactory("fishtest")
    config = Configurator(
        settings=settings,
        session_factory=session_factory,
        root_factory="fishtest.models.RootFactory",
    )
    config.include("pyramid_mako")
    config.set_default_csrf_options(require_csrf=False)
```

**Comment by vdbergh** [comment]
> What do you mean it does not save state on the server, then how we authenticate web views? 
You do it in the way  that I explained. Note that if you save state on the server you need a mechanism to make it survive server restarts.

We already have a session management, I assume they are saved here
> 
> ```
>     """This function returns a Pyramid WSGI application."""
>     session_factory = SignedCookieSessionFactory("fishtest")
>     config = Configurator(
>         settings=settings,
>         session_factory=session_factory,
>         root_factory="fishtest.models.RootFactory",
>     )
>     config.include("pyramid_mako")
>     config.set_default_csrf_options(require_csrf=False)
> ```

**Comment by vdbergh** [comment]
You can do it without changing server state in the way I explained. I have the impression that this is how pyramid does it under the hood.

Note that if you change server state you need a mechanism to make it survive server restarts.

**Comment by peregrineshahin** [comment]
Please note that a session based authentication by design does not want to survive a server restart, sessions mean that we can expire the session and easily generate a new one if the token is not used.

We even remove the token when a user logout.

**Comment by vdbergh** [comment]
> Please note that a session based authentication by design does not want to survive a server restart, sessions mean that we can expire the session and easily generate a new one if the token is not used.
> 
> We even remove the token when a user logout.

Are you sure the authenticated views to not survive server restart?

**Comment by vdbergh** [comment]
Here is an explanation of pyramid authentication:

https://stackoverflow.com/questions/12765349/pyramid-authtktauthenticationpolicy-secret-parameter

**Comment by peregrineshahin** [comment]
Here is my CSRF-token, the server saves it of course, and if I keep refreshing I get the same token
Now if I logout and login I will get different token
![Capture](https://github.com/glinscott/fishtest/assets/41402573/eaec261e-06e6-4a10-bf72-38855730e8bc)

**Comment by vdbergh** [comment]
Why need the server save it? It can just return it (or regenerate it).

**Comment by vdbergh** [comment]
That views would not survive server restart would be rather inconvenient. It would mean that each time the server is restarted the user needs to log in again.

**Comment by vdbergh** [comment]
Anyway you have not explained what is wrong with the scheme I proposed (which I think is what pyramid uses) and which does not require changing any server state (so it is trivial to implement).

**Comment by peregrineshahin** [comment]
I'm not sure you understand my solution. and I'm not sure I understand your solution either.

**Comment by vdbergh** [comment]
I understand your solution perfectly (just have a data structure on the server with user/token/timestamp). 

Here is the stateless solution (I will simplify it as much as I can, e.g. no expiry timestamps/ip addresses etc):

1. User logs in with username/name password.

2. If password is correct then the server sends back the signature of the username (uses ~/fishtest.secret) .

3. Next time the user presents username+signature.  The server verifies the signature (uses ~/fishstest.secret). If the signature checks out then the user is considered being logged in.

**Comment by peregrineshahin** [comment]
Well, IIUC then the solution you are presenting is using a hashing algorithm that with a string gives each time the same signiture?
couple of things to take in mind is:
1 - with a session based you can use a very cheap string instead of a complex hash algorithm
2 - with a non-session based the signature always work, thus if someone got his key exposed then there is no way to update new keys unless we change the secret string for the whole server and thus make the saved signature not valid for all other users. (in other words it's the same as having another password)
where as if someone know your CSRF token all what you can do is just logout and then the CSRF token is no longer valid.

**Comment by ppigazzini** [comment]
I made these tests on DEV, hoping that it would be helpful:
1. logged as user01, created a test, copied the `crsf-token` from the html page, kept the page open
2. restarted the fishtest server `sudo systemctl restart fishtest@{6543..6545}`
3. created successfully a new test, `crsf-token` not changed
4. restarted nginx `sudo systemctl restart nginx`
5. created successfully a new test, `crsf-token` not changed
6. rebuilt fishtest server from scratch (old folder deleted)
7. created successfully a new test, `crsf-token` not changed

In recent years, I frequently encountered an `invalid CSRF` message while testing on DEV. Logging out and back in resolved the issue. However, I haven’t seen that error message in the last few months.

**Comment by vdbergh** [comment]
@ppigazzini Thanks! So this is consistent with my theory that the token is computed rather than randomly generated.

Did you preserve ~/fishtest.secret while reinstalling the server?

**Comment by ppigazzini** [comment]
https://github.com/glinscott/fishtest/blob/02c55acc9c1ed4189795b140bf3ac76738a90f77/server/fishtest/__init__.py#L60-L65

Yes

**Comment by vdbergh** [comment]
Could change it (of course preserving it)? That should change the token.

**Comment by ppigazzini** [comment]
1. duplicated the browser tab with the `crsf-token`
2. changed `fishtest.secret` and restarted the server
3. tab-0 requested a login to add a new test
4. restored the original `fishtest.secret` and restarted the server
5. tab-1 added successfully a new test

**Comment by vdbergh** [comment]
@ppigazzini  Thanks again! So I think that authentication in pyramid is indeed stateless. The only state on the server is the file `~/fishtest.secret`.

**Comment by ppigazzini** [comment]
pyramid tutorial about users authentication (with password encryption as well).
https://docs.pylonsproject.org/projects/pyramid/en/latest/quick_tutorial/authentication.html

**Comment by vdbergh** [comment]
So the `csrf-token` (which is not a cookie) is actually to prevent cross site request forgery attacks. It is not directly responsible for authentication. The authentication is done by the cookie `auth_tkt` (valid for one year if the user wants to be remembered). 

It is a bit suprising that the csrf-token seems to be constant in @ppigazzini's experiments. There is something I do not understand yet.

**Comment by peregrineshahin** [comment]
I never said CSRF_token is responsible for the authentication. authentication means you login and only login. CSRF_token helps in WEB to verifiy you are the one who makes the sub requests which is what matters and not the initial authentication step, I don't know how is it generated with Pyramid we can have a normal/random token we implement ourselfs.

So the concept is a simple session based token that is CSRF-alike. I apologies I assumed how pyramid does it.

**Comment by ppigazzini** [comment]
> It is a bit suprising that the csrf-token seems to be constant in @ppigazzini's experiments. There is something I do not understand yet.

The `crsf-token` changes after a logout.
The `crsf-token` doesn't change after a login.

**Comment by vdbergh** [comment]
> > > It is a bit suprising that the csrf-token seems to be constant in @ppigazzini's experiments. There is something I do not understand yet.
> > 
> > 
> > The `crsf-token` changes after an user logout. The `crsf-token` doesn't change after an user login.

Yes but the surprising thing is that it is also preserved after server restart. This suggests that this state is kept on the client as well...
> 
> Yes which means there is of course a state .. it's not remotely stateless as vdbergh claim.. weather the state is saved in the client or the server is another question

We are of course talking about server state. All server state is a headache (restarts/multiple threads/instances etc...).

It seems to me that the csrf scenario does not apply to the worker/server context (I do not see what a cross site request forgery would mean in that context). It is really about stateless vs stateful authentication.  I must say I very much prefer the stateless scenario.

**Comment by peregrineshahin** [comment]
Take this cool test
1. login
2. create a test
3. go remove cookies
4. try to stop the test

it will not work, that means the session validity is saved in the client.

I prefer the state solution because simply we can block workers from getting task in the future;)

**Comment by ppigazzini** [comment]
Perhasp it depends upon this code:
https://github.com/glinscott/fishtest/blob/02c55acc9c1ed4189795b140bf3ac76738a90f77/server/fishtest/views.py#L212-L217

**Comment by peregrineshahin** [comment]
> I prefer the state solution because simply we can block workers from getting task in the future ;)
This was requested by long time contributer if their worker is broken

**Comment by peregrineshahin** [comment]
Well if we are going by a stateless solution that is not a session based, I actually don't see how is that different from caching a map with password, hashed password to be honest.
on the other hand, if we have state then we can invalidate a session if the user changed the password.
The key will always work == the user have another kind of password (i.e. no advantage security wise over that solution) it's extra step of hashing also on initial login (the hashing of the login and the hashing to make the key).

**Comment by vdbergh** [comment]
Obviously blocking a user/worker needs some server state.  But from my point of view this is a different problem from authentication.

**Comment by ppigazzini** [comment]
I wonder if the worker heartbeat needs authentication.

**Comment by vdbergh** [comment]
> Suppose someone say to you my password got hacked and there is someone who uses their worker name sending bad results to fishtest. The first thing the user will do is to change his password thinking that the other person will gets disconnected immediatly.
> 
> It's a bummer if it's not the case.

This is only a problem if the client side token also gets stolen (the attacker needs it for the password less login). So it seems best to not store this token on disk. Of course the client side token will also have an expiration timestamp, after which it will have to be refreshed.

I am thinking this is also an issue with the views implementation in pyramid. If an attacker has full control over the client's browser and steals all credential related data then it seems to me that the server cannot block the attacker until something expires.

**Comment by peregrineshahin** [comment]
I'm not sure you realize that login in with password should always be an option in all cases except if the worker want to login via a token
Or else how would the user login from another place if you know the password we will send the token of course..

The way one can guard against stolen password then is simply by forgetting it or changing it thus the generated token must have the password as input to make it invalid.

Current master if I changed my password then all the workers connected should not be valid to send anything

**Comment by peregrineshahin** [comment]
That's only if the token is not random. And saved on the server of course then you can easily invalidate the random token via invalidating the session.

**Comment by peregrineshahin** [comment]
Once the user change password.

**Comment by peregrineshahin** [comment]
The way pyramid does it is give an csrf-token to each session not each user.
I think it would be an overkill.

forms can send cookies when submitting data to the server. When a form is submitted, the browser includes the cookies associated with the domain in the HTTP request, including both GET and POST requests.

The cookies are automatically sent by the browser as part of the HTTP request headers, specifically in the "Cookie" header. This allows the server to receive and process the cookies along with the form data.

**Comment by peregrineshahin** [comment]
but then There should be a difference between signing out from worker stuff and killing a worker then
idk, what was the problem again with plain password? :Kappa:

**Comment by vdbergh** [comment]
> I'm not sure you realize that login in with password should always be an option in all cases except if the worker want to login via a token Or else how would the user login from another place if you know the password we will send the token of course..
> 
> The way one can guard against stolen password then is simply by forgetting it or changing it thus the generated token must have the password as input to make it invalid.
> 
> Current master if I changed my password then all the workers connected should not be valid to send anything

I was just commenting on your claim that with a stateless token one cannot block a user (other than waiting for the token to time out). I was saying that this is only true if the attacker can also steal the token (besides the password). If the token is not saved on disk this is rather unlikely.

You do not have to explain to me how a stateful token works. I understand it perfectly.

**Comment by peregrineshahin** [comment]
Well we use SignedCookieSessionFactory for views (i.e. a Signed cookie that is saved on the client side).

"A signed token is a good method for anything where you want to issue a token and then, when it is returned, be able to verify that you issued the token, without having to store any data on the server side."

I must say it's pretty hard for the browser to reveal for a third party what the cookies are to begin with, by current standards.
But if the client tried hard enough they might view the content of the session but will not be able to change it because of the hmac signiture.

So you are correct there is no way we can invalidate cookies we issued via this method without a storage, I don't know of any big websites that don't know who are logged in on their site worth noting that using a refresh token and an access token might mitigate the issue more or less.

So, while big websites may have token storage mechanisms in place, they still strive to maintain a stateless architecture by using short-lived tokens and stateless token validation techniques.

**Comment by vdbergh** [comment]
But if I steal you password I can just use it to log in to get a new token, before you had the chance to change your password. As I see it, this token will then continue to work until it expires even after you change your password (I am assuming a stateless server).

**Comment by peregrineshahin** [comment]
> But if I steal you password I can just use it to log in to get a new token, before you had the chance to change your password. As I see it, this token will then continue to work until it expires even after you change your password (I am assuming a stateless server).

And the answer is yes. that's why I said "So you are correct there is no way we can invalidate cookies we issued via this method without a storage"

**Comment by peregrineshahin** [comment]
This concludes IMO that the solution should be Server-based stateful system and it is appreciated if we can find the same thing for Views.

I'm pretty sure Modern WEB Frameworks have that option or can integrate it handily at least via some dependencies or custom work.

**Comment by vdbergh** [comment]
It seems the minimum amount of state required on the server is one timestamp. Which can be used to expire all client side tokens.

**Comment by peregrineshahin** [comment]
> It seems the minimum amount of state required on the server is one timestamp. Which can be used to expire all client side tokens.

It would be a nice idea if we don't overload the server with logins and expire users tokens at the same time.

**Comment by peregrineshahin** [comment]
> > It seems the minimum amount of state required on the server is one timestamp. Which can be used to expire all client side tokens.
> 
> It would be a nice idea if we don't overload the server with logins and expire users tokens at the same time.

Thus a change of user credentials should not affect another user token not to be valid.

**Comment by vdbergh** [comment]
> > It seems the minimum amount of state required on the server is one timestamp. Which can be used to expire all client side tokens.
> 
> It would be a nice idea if we don't overload the server with logins and expire users tokens at the same time.

Well password change happens very rarely. So this would not be a big deal. And in any case the workers have to re-login after server restarts, which happen frequently.

**Comment by vdbergh** [comment]
Anyway this discussion is mostly theoretical since it seems both solutions are feasible. And as Vondele points out, with the memoizing of the hash function, there is actually no pressing need to implement either of them. But I learnt something since I did not know about stateless tokens before.
